src/codeeditor/editor.ts

```diff
@@ -0,0 +1,804 @@
+   */
+  export
+  interface ICursorPosition {
+    /**
+     * The cursor line number.
+     */
+    line: number;
+
+    /**
+     * The cursor column number.
+     */
+    column: number;
+  }
+
+  /**
+   * An interface describing editor state coordinates.
+   */
+  export
+  interface ICoords {
+    /**
+     * The left coordinate value.
+     */
+    readonly left: number;
+
+    /**
+     * The right coordinate value.
+     */
+    readonly right: number;
+
+    /**
+     * The top coordinate value.
+     */
+    readonly top: number;
+
+    /**
+     * The bottom coordinate value.
+     */
+    readonly bottom: number;
+  }
+
+  /**
+   * Configuration options for the editor.
+   */
+  export
+  interface IConfiguration extends IDisposable {
+    /**
+     * A signal emitted when the configuration changes.
+     */
+    changed: ISignal<IConfiguration, string>;
+
+    /**
+     * Whether line numbers should be displayed. Defaults to false.
+     */
+    lineNumbers: boolean;
+
+    /**
+     * Set to false for horizontal scrolling. Defaults to true.
+     */
+    wordWrap: boolean;
+
+    /**
+     * Whether the editor is read-only.  Defaults to false.
+     */
+    readOnly: boolean;
+
+    /**
+     * The number of spaces a tab is considered equal to. Defaults to 4.
+     */
+    tabSize: number;
+
+    /**
+     * Translate tabs to spaces. Defaults to true.
+     */
+    tabsToSpaces: boolean;
+
+    /**
+     * Set to false to stop auto pairing quotes, brackets etc.
+     * Defaults to true.
+     */
+    autoMatchEnabled: boolean;
+
+    /**
+     * Set to false to disable highlighting the brackets surrounding the
+     * cursor. Defaults to true.
+     */
+    matchBrackets: boolean;
+
+    /**
+     * Set to false to disable scrolling past the end of the buffer.
+     * Defaults to false.
+     */
+    scrollPastEnd: boolean;
+  }
+
+  /**
+   * A text selection.
+   */
+  export
+  interface ITextSelection {
+    /**
+     * The index to the first character in the current selection.
+     */
+    start: number;
+
+    /**
+     * The index to the last character in the current selection.
+     */
+    end: number;
+
+    /**
+     * The direction in which selection occurred.
+     */
+    direction: 'forward' | 'backward' | 'none';
+  }
+
+  /**
+   * An editor model.
+   */
+  export
+  interface IModel extends IDisposable {
+    /**
+     * A signal emitted when the value changes.
+     */
+    valueChanged: ISignal<IModel, IChangedArgs<string>>;
+
+    /**
+     * A signal emitted when a property changes.
+     */
+    propertyChanged: ISignal<IModel, IChangedArgs<any>>;
+
+    /**
+     * The text stored in the model.
+     */
+    value: string;  // TODO: this should be an iobservablestring.
+
+    /**
+     * The uri associated with the model.
+     */
+    uri: string;
+
+    /**
+     * A mime type of the model.
+     */
+    mimeType: string;
+
+    /**
+     * The cursor position.
+     */
+    cursor: ICursorPosition;
```

Comment:
In CodeMirror these are in the Doc, which I'm equating with `IModel`, which I'd consider to be a view model.  This makes it so that changes are set on the view model programmatically and on the widget by user interaction, avoiding things like #682.